sql, distsql: support DISTINCT ON clause in SELECT

richardwu opened the pull request:
Meta note: I still need to add more DistSQL test cases and a few more normal test cases, but wanted to push this out ASAP to get some feedback. 

**A big change from the previous proposed strategy:** instead of mangling with group bys and the `FIRST` aggregator, it appears that it is much cleaner and more straightforward to adapt the existing `DISTINCT` logic.

Closes #12633 

Release notes (feature): the DISTINCT ON clause was introduced for
SELECT statements.

Syntax
```sql
SELECT DISTINCT ON (<distinct-on-exprs>) <select-exprs> FROM x ...
```

Which will retrieve an arbitrary row of <select-exprs> for every distinct
combination of <distinct-on-exprs>.

If the ORDER BY clause is specified in the SELECT statement, DISTINCT ON
will respect that ordering when choosing the "arbitrary row" (that is:
it will choose to first row based on the ordering).

For more examples, see `logic_test/distinct_on` and `logic_test/distsql_distinct_on`.

cc: @RaduBerinde

richardwu committed:
sql, distsql: support DISTINCT ON clause in SELECT

Release notes (feature): the DISTINCT ON clause was introduced for
SELECT statements.

Syntax
```sql
SELECT DISTINCT ON (<distinct-on-exprs>) <select-exprs> FROM x ...
```

Which will retrieve an arbitrary row of <select-exprs> for every distinct
combination of <distinct-on-exprs>.

If the ORDER BY clause is specified in the SELECT statement, DISTINCT ON
will respect that ordering when choosing the "arbitrary row" (that is:
it will choose to first row based on the ordering).

This patch also propagates weak keys for DISTINCT node in general.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20463)
<!-- Reviewable:end -->

RaduBerinde commented:
Review status: 0 of 14 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsql_physical_planner.go, line 1966 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cLLZ4CjR_wU0nVwav:-L-cLLZ4CjR_wU0nVwaw:b-vcb8qz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/distsql_physical_planner.go#L1966)):*
> ```Go
> 
> 	// In cases like:
> 	//   SELECT a FROM t ORDER BY b
> ```

This should be an example that is relevant to this code (something with DISTINCT)

---

*[pkg/sql/distsql_physical_planner.go, line 1967 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cLS9scvKkiFuvwmBt:-L-cLS9scvKkiFuvwmBu:bnbryzc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/distsql_physical_planner.go#L1967)):*
> ```Go
> 	// In cases like:
> 	//   SELECT a FROM t ORDER BY b
> 	// we have columns (b) that are only used for sorting. These columns
> ```

only used for DISTINCT ON

---

*[pkg/sql/distsql_physical_planner.go, line 1968 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cLUXVU8tLI4uHgXRR:-L-cLUXVU8tLI4uHgXRS:bu960f7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/distsql_physical_planner.go#L1968)):*
> ```Go
> 	//   SELECT a FROM t ORDER BY b
> 	// we have columns (b) that are only used for sorting. These columns
> 	// are not in the output columns of the sortNode; we set a projection
> ```

distinctNode

---

*[pkg/sql/distsql_physical_planner.go, line 1977 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cLDvLRfr0V9z7sXSz:-L-cLDvLRfr0V9z7sXT-:bc46ujj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/distsql_physical_planner.go#L1977)):*
> ```Go
> 		outputCols := make([]uint32, len(n.columns))
> 		for i, col := range plan.planToStreamColMap {
> 			outputCols[i] = uint32(col)
> ```

This code should check for -1, see #20489

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
In plan_physical_props.go, we have
```
	case *distinctNode:
		return planPhysicalProps(n.plan)
```

This needs to be adjusted. The distinctNode can project away columns; so the physical props of the child can't be used as is. See `physicalProps.project()`

I'd be curious what happens with the current change for something like `SELECT DISTINCT ON (a, c) a, b FROM abc` where a, b, c is a primary key.

---

Review status: 0 of 14 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distinct.go, line 55 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cMykH19dCIzsncYmC:-L-cMykH19dCIzsncYmD:bp8lls3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/distinct.go#L55)):*
> ```Go
> 		return nil, nil
> 	}
> ```

[nit] `d := &distinctNode{columns: r.columns}` can be used for both cases below

---

*[pkg/sql/expand_plan.go, line 571 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cNW9ZB0HrSsCbJFT-:-L-cNW9_8fXZS0WS_Wsl:brx60ex) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/expand_plan.go#L571)):*
> ```Go
> 
> 	case *distinctNode:
> 		// distinctNode uses whatever order the underlying node presents (regardless
> ```

This is a relatively minor improvement, but we should only use whatever prefix of the source ordering is useful for the DISTINCT ON columns. For example

`SELECT a, b DISTINCT ON (a) FROM ab` where a, b is the primary key. The scanNode will have ordering `a+, b+` but the `b+` is not useful. Simplifying is desirable to reduce the number of columns in ordered synchronizers.

---

*[pkg/sql/render.go, line 199 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cMKHauzGZsRs07IJc:-L-cMKHauzGZsRs07IJd:b-nwkg34) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/render.go#L199)):*
> ```Go
> 	// NB: orderBy, window, and groupBy are passed and can modify the
> 	// renderNode, but must do so in that order.
> 	// Note this order is exactly the reverse of the order of how the plan
> ```

Note that

---

*[pkg/sql/render.go, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cMGVQtAW9uu--l4G3:-L-cMGVQtAW9uu--l4G4:b-k70meb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/render.go#L206)):*
> ```Go
> 	//	 |
> 	//	 v
> 	//    sortNode
> ```

[nit] alignment

---

*[pkg/sql/logictest/testdata/logic_test/distinct_on, line 4 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cPSfRio1gjjTjp0dc:-L-cPSfRio1gjjTjp0dd:bis9q7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/logictest/testdata/logic_test/distinct_on#L4)):*
> ```
> 
> statement ok
> CREATE TABLE xyz (
> ```

would be useful to have some PK columns and have queries with PK columns, or non-PK-columns, or combinations (to verify computed Orderings).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-:-L-cOfsWRaYxO7bUYt3z:bsr8ce3)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Ahhh good catch. Looks like it doesn't produce the correct plan for
```
CREATE TABLE abc (
  a STRING,
  b STRING,
  c STRING,
  PRIMARY KEY (a, b, c)
)

INSERT INTO abc VALUES
  ('1', '1', '1'),
  ('1', '1', '2'),
  ('1', '2', '2')

SELECT DISTINCT ON (a, c) a, b FROM abc
```

Before

```
0  scan  ·      ·            (a, b, c)  a!=NULL; b!=NULL; c!=NULL; key(a,b,c)
0  ·     table  abc@primary  ·          ·
0  ·     spans  ALL          ·          ·
```
and executing that query throws an `index out of range` error 
```
panic: runtime error: index out of range

goroutine 1483 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*distSQLReceiver).Push(0xc42a837200, 0xc42915fc30, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/richardwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:373 +0x9f2
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcOutputHelper).EmitRow(0xc42915fc00, 0x652e840, 0xc42934f680, 0xc42a760068, 0x2, 0x2, 0xc42950bdd8, 0x1, 0x101)
        /Users/richardwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:275 +0x161
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.emitHelper(0x652e840, 0xc42934f680, 0xc42915fc00, 0xc42a760068, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/richardwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:218 +0x3ce
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*noopProcessor).Run(0xc42915fc00, 0x652e840, 0xc42934f680, 0xc42a5307c8)
        /Users/richardwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:416 +0x231
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
        /Users/richardwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:419 +0x3d7
```

After
```
0  render    ·            ·            (a, b)     a!=NULL; b!=NULL; +a,+b
0  ·         render 0     a            ·          ·
0  ·         render 1     b            ·          ·
1  distinct  ·            ·            (a, b, c)  a!=NULL; b!=NULL; c!=NULL; key(a,b,c); +a,+b,+c
1  ·         distinct on  a, c         ·          ·
1  ·         order key    a            ·          ·
2  scan      ·            ·            (a, b, c)  a!=NULL; b!=NULL; c!=NULL; key(a,b,c); +a,+b,+c
2  ·         table        abc@primary  ·          ·
2  ·         spans        ALL          ·          ·
```
with results
```
1 1
1 1
```

So I've taken @knz suggestion that `distinctNode` should not do any projection (but rather, we impose a `postRender renderNode` if `DISTINCT ON` ends up adding additional render expressions/columns). 

It's important we do this because a `distinctNode` [may be elided if its expressions form a strong key](https://github.com/cockroachdb/cockroach/blob/6f6c3bf7699e876e6802f9e20fc5047c5c84a79c/pkg/sql/expand_plan.go#L202-L208). We could attempt to splice in a `renderNode` when the `distinctNode` is elided, but that becomes difficult since we no longer have access to the `SelectClause.Exprs` during the `expand_plan` phase. We could of course memoize those original expressions in the `distinctNode`, but that, in my opinion, violates separation of concerns.



Also, I realized that the size of this PR has blown up from its initial size: I don't see a logical way to split this up since 900 lines of it are logic tests. 

I apologize in advance to all reviewers: let me know if you want me to somehow break this up into multiple pieces.

---

Review status: 0 of 18 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distinct.go, line 55 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cMykH19dCIzsncYmC:-L-rs7SMQ9exdDISIBcM:b59c09r) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/distinct.go#L55)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `d := &distinctNode{columns: r.columns}` can be used for both cases below
</blockquote></details>

Done (removed `columns` field in favor of a `postRender renderNode` for projecting expressions).

---

*[pkg/sql/distsql_physical_planner.go, line 1966 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cLLZ4CjR_wU0nVwav:-L-lnhIZgiHMz1m-X1Cs:b-pj6uyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/distsql_physical_planner.go#L1966)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This should be an example that is relevant to this code (something with DISTINCT)
</blockquote></details>

Fixed this comment for DISTINCT, thanks.

---

*[pkg/sql/distsql_physical_planner.go, line 1977 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cLDvLRfr0V9z7sXSz:-L-rsJ2CJcpFIfIohW2p:b2tklb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/distsql_physical_planner.go#L1977)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This code should check for -1, see #20489
</blockquote></details>

No longer required since we're no longer projecting within the `distinctNode` (see the top-level comment above).

---

*[pkg/sql/expand_plan.go, line 571 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cNW9ZB0HrSsCbJFT-:-L-s-lL_gLhquv6GZo_o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/expand_plan.go#L571)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is a relatively minor improvement, but we should only use whatever prefix of the source ordering is useful for the DISTINCT ON columns. For example

`SELECT a, b DISTINCT ON (a) FROM ab` where a, b is the primary key. The scanNode will have ordering `a+, b+` but the `b+` is not useful. Simplifying is desirable to reduce the number of columns in ordered synchronizers.
</blockquote></details>

Done.

---

*[pkg/sql/render.go, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cMGVQtAW9uu--l4G3:-L-s0DFVWbnrNAjoBAe5:bql6urd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/render.go#L206)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] alignment
</blockquote></details>

Should be fixed now.

---

*[pkg/sql/logictest/testdata/logic_test/distinct_on, line 4 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L-cPSfRio1gjjTjp0dc:-L-s0FqHVeucGhrQZkND:b-cha5n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2952cb543d65d0315fd93c49c41196d1acd02085/pkg/sql/logictest/testdata/logic_test/distinct_on#L4)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

would be useful to have some PK columns and have queries with PK columns, or non-PK-columns, or combinations (to verify computed Orderings).
</blockquote></details>

Updated with a PK table and added 2 PK columns to xyz.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-:-L-i3htSjMXOzEt5xPL1:b-c3xjtq)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Yes, using a `renderNode` for the projection is great.

---

Review status: 0 of 19 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distinct.go, line 72 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05qra0n9JZWawVHvHQ:-L05qra1Rs0wOhEiLKL2:b7gvqjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/distinct.go#L72)):*
> ```Go
> 
> 	for _, expr := range n.DistinctOn {
> 		// The expression in DISTINCT ON follow similar rules as
> ```

[nit] expressions (or follows)

---

*[pkg/sql/distinct.go, line 335 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05vRERNoQ2dXgqTswA:-L05vRERNoQ2dXgqTswB:bl2y84x) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/distinct.go#L335)):*
> ```Go
> 	}
> 
> 	return props
> ```

We should add the set of ON columns as a weak key.

---

*[pkg/sql/expand_plan.go, line 285 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05ro8XDV0R6Znd0l4k:-L05ro8XDV0R6Znd0l4l:b-gl4jpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/expand_plan.go#L285)):*
> ```Go
> 	distinctPp := d.computePhysicalProps()
> 
> 	if !distinctPp.isEmpty() {
> ```

NYC but `isEmpty` is broken and/or badly named, it doesn't check if we have weak keys (plus, we may have equivalency groups). I would remove the call here (ideally, we should rename or remove the function).

---

*[pkg/sql/expand_plan.go, line 588 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05v8xZhZkSCqldqkfQ:-L05v8x_MLVNFU1q6BBZ:b-k0i1la) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/expand_plan.go#L588)):*
> ```Go
> 		// applicable).
> 		// It is not necessary to maintain ordering of other columns
> 		// unless explicitly required by the child node (i.e.
> ```

I don't understand the "unless explicitly required by the child node" part.

---

*[pkg/sql/expand_plan.go, line 590 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05uqGXU3mtzWR3Of5F:-L05uqGXU3mtzWR3Of5G:b-9n8zgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/expand_plan.go#L590)):*
> ```Go
> 		// unless explicitly required by the child node (i.e.
> 		// sortNode).
> 		// This improves performance since ordering incurs an overhead
> ```

This is not necessary, it's already explained in the simplifyOrderings comment.

---

*[pkg/sql/expand_plan.go, line 592 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05yhNpq2aItWc1OLfH:-L05yhNpq2aItWc1OLfI:b-k5cyuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/expand_plan.go#L592)):*
> ```Go
> 		// This improves performance since ordering incurs an overhead
> 		// when merging streams in distsql.
> 		sourceOrdering := n.computePhysicalProps()
> ```

I am not sure that folding this into computePhysicalProps is the right thing to do; there's a reason why simplifyOrderings is a separate pass: we want the full orderings during expand plan in case we can take advantage of something.

For example
`SELECT k, v DISTINCT ON (v)  ORDER BY k`
We want to elide the sortNode here; we can't if we remove the ordering. I'd be curious what the output of `EXPLAIN (VERBOSE)` is for this.

The simplifyOrderings pass should simplify the orderings as much as possible while still satisfying `usefulOrdering` (but no more!)

---

*[pkg/sql/render.go, line 785 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05vn-OdH2xGTIht02M:-L05vn-OdH2xGTIht02N:bw9wt1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/render.go#L785)):*
> ```Go
> 				if col.Name == target {
> 					if index != -1 {
> 						// There is more than one
> ```

[nit] reformat this comment (and the one below) to 80 cols

---

*[pkg/sql/logictest/testdata/logic_test/distsql_distinct_on, line 146 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05wDs7IPZK2RNkSAh3:-L05wDs7IPZK2RNkSAh4:b-by0844) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/logictest/testdata/logic_test/distsql_distinct_on#L146)):*
> ```
> 
> query T
> SELECT "URL" FROM [EXPLAIN (DISTSQL) SELECT DISTINCT ON (a, b) a, b FROM abc ORDER BY a, b, c]
> ```

would be useful to have the `EXPLAIN (VERBOSE)` for these queries as well, to see orderings and other properties.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-:-L05qUt4jGAWFTN7S3uv:bfx03v7)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Review status: 0 of 21 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/distinct.go, line 72 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05qra0n9JZWawVHvHQ:-L06eUwRxCciShj9PUxy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/distinct.go#L72)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] expressions (or follows)
</blockquote></details>

Done.

---

*[pkg/sql/distinct.go, line 335 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05vRERNoQ2dXgqTswA:-L06eV_Jg8bcAjLX2ABf:b-mdg4o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/distinct.go#L335)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should add the set of ON columns as a weak key.
</blockquote></details>

So I've renamed this method to `projectChildPropsToOnExprs` (which is used in `expandPlan` to figure out if the `distinctNode` can be elided and to compute `usefulOrdering` for optimizing the distinct process). The `physicalProps` returned by this method is the pre-distinct `physicalProps` for just the `ON` expressions.

The actual `physicalProps` guaranteed by the `distinctNode` would be the child plan's `physicalProps` with the added weak keys. I've updated the code in `planPhysicalProps`. I believe this is how everything should fit together, but correct me if I'm wrong.

---

*[pkg/sql/expand_plan.go, line 285 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05ro8XDV0R6Znd0l4k:-L06fE9byFR2MeHpZSfm:b-dgenjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/expand_plan.go#L285)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

NYC but `isEmpty` is broken and/or badly named, it doesn't check if we have weak keys (plus, we may have equivalency groups). I would remove the call here (ideally, we should rename or remove the function).
</blockquote></details>

Ah I didn't even realize this only checked ordering! I removed it (I'll open a PR to refactor this).

---

*[pkg/sql/expand_plan.go, line 588 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05v8xZhZkSCqldqkfQ:-L068v6QNvptQY5tpgCb:bzezgc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/expand_plan.go#L588)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't understand the "unless explicitly required by the child node" part.
</blockquote></details>

I guess what I was trying to say is the child plan can ignore `usefulOrdering` in the recursive call (in the case of `sortNode` where `needSort = true`. This comment doesn't really belong here so I've removed it (and updated it to a more a relevant note).

---

*[pkg/sql/expand_plan.go, line 590 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05uqGXU3mtzWR3Of5F:-L069BG5VJ6-XzENfXxh:b1tww5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/expand_plan.go#L590)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is not necessary, it's already explained in the simplifyOrderings comment.
</blockquote></details>

Agreed :+1: s

---

*[pkg/sql/expand_plan.go, line 592 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05yhNpq2aItWc1OLfH:-L06Z2ghI3cPzaf6aRk7:b-19n3qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/expand_plan.go#L592)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I am not sure that folding this into computePhysicalProps is the right thing to do; there's a reason why simplifyOrderings is a separate pass: we want the full orderings during expand plan in case we can take advantage of something.

For example
`SELECT k, v DISTINCT ON (v)  ORDER BY k`
We want to elide the sortNode here; we can't if we remove the ordering. I'd be curious what the output of `EXPLAIN (VERBOSE)` is for this.

The simplifyOrderings pass should simplify the orderings as much as possible while still satisfying `usefulOrdering` (but no more!)
</blockquote></details>

So it's required that the `DISTINCT ON` expression be a [prefix or fully contain](https://github.com/cockroachdb/cockroach/pull/20463/files#diff-143ec11b909537e265bd82b914b05335R227) the `ORDER BY` clause attached to it.

That is: the ordering that is "useful" to (or required? by) a `DISTINCT ON` node will always be a subset of its `ON` expressions.

---

*[pkg/sql/render.go, line 785 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05vn-OdH2xGTIht02M:-L06d1Ep4fxg_ha1OVPZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/render.go#L785)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] reformat this comment (and the one below) to 80 cols
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/distsql_distinct_on, line 146 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05wDs7IPZK2RNkSAh3:-L06eRauwuOpdRhsue8x:b-npjcwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/logictest/testdata/logic_test/distsql_distinct_on#L146)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

would be useful to have the `EXPLAIN (VERBOSE)` for these queries as well, to see orderings and other properties.
</blockquote></details>

Good point, added!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 21 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distinct.go, line 338 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0BAIR6DwbIMJCGOLam:-L0BAIR6DwbIMJCGOLan:b-9ola5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72d9cf0bacccf752dbe5de9fc367614c857f460/pkg/sql/distinct.go#L338)):*
> ```Go
> 			}
> 		}
> 		return props.project(colMap)
> ```

I would add here a comment along the lines of "The DISTINCT ON columns are the same or a prefix of the ORDER BY columns; in addition, the ON columns form a key, so any extra columns in the underlying ordering don't help".

---

*[pkg/sql/expand_plan.go, line 592 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L05yhNpq2aItWc1OLfH:-L0B9gdWhRB3dsKWgHqT:bo50k8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/477d9933470cb4d681e5a60ff13d5787fca51a74/pkg/sql/expand_plan.go#L592)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

So it's required that the `DISTINCT ON` expression be a [prefix or fully contain](https://github.com/cockroachdb/cockroach/pull/20463/files#diff-143ec11b909537e265bd82b914b05335R227) the `ORDER BY` clause attached to it.

That is: the ordering that is "useful" to (or required? by) a `DISTINCT ON` node will always be a subset of its `ON` expressions.
</blockquote></details>

Oh! That makes sense.

---

*[pkg/sql/plan_physical_props.go, line 114 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0BB2YqE2fcF2Rx0WgW:-L0BB2YqE2fcF2Rx0WgX:bvxdrt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72d9cf0bacccf752dbe5de9fc367614c857f460/pkg/sql/plan_physical_props.go#L114)):*
> ```Go
> 	// addWeakKey will replace an existing superset weak key with the
> 	// specified weak key.
> 	// Since distinct does not eliminate keys (it can only produce an
> ```

We don't need to do this. addWeakKey already looks at existing groups to make sure there are no redundancies.

It's also doesn't seem correct. We want the smaller set to be the new key (a smaller set is a stronger guarantee; it implies that any super-set is also a key). I would add a test for this, e.g. `SELECT DISTINCT ON(a) a, b` where `a,b` are a PK (we want `a` to be the key, not `a,b`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Review status: 0 of 24 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/distinct.go, line 338 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0BAIR6DwbIMJCGOLam:-L0BEGrtkiy8aaTI0vWk:bp98sir) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72d9cf0bacccf752dbe5de9fc367614c857f460/pkg/sql/distinct.go#L338)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I would add here a comment along the lines of "The DISTINCT ON columns are the same or a prefix of the ORDER BY columns; in addition, the ON columns form a key, so any extra columns in the underlying ordering don't help".
</blockquote></details>

:+1: done

---

*[pkg/sql/plan_physical_props.go, line 114 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0BB2YqE2fcF2Rx0WgW:-L0BDP8h5V-pYslSaUGZ:bk1kmx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72d9cf0bacccf752dbe5de9fc367614c857f460/pkg/sql/plan_physical_props.go#L114)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We don't need to do this. addWeakKey already looks at existing groups to make sure there are no redundancies.

It's also doesn't seem correct. We want the smaller set to be the new key (a smaller set is a stronger guarantee; it implies that any super-set is also a key). I would add a test for this, e.g. `SELECT DISTINCT ON(a) a, b` where `a,b` are a PK (we want `a` to be the key, not `a,b`).
</blockquote></details>

Ah gotcha, for some reason I thought more columns in a key was "better". Thanks for the insight! 

Added the test in `testlogic/distinct_on`.

---

*[pkg/sql/plan_physical_props.go, line 121 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0BIAW_jzUvuQp_ZQvs:-L0BIAW_jzUvuQp_ZQvt:b-yrfvqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffec33995ca68a379b87ac8e6a87e78a149dcfcd/pkg/sql/plan_physical_props.go#L121)):*
> ```Go
> 		// form a key after distinct.
> 		var allCols util.FastIntSet
> 		allCols.AddRange(0, len(planMutableColumns(n.plan))-1)
> ```

Also added weak keys for full `DISTINCT` (no `ON` clause), which it seems we weren't propagating before.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 24 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distinct.go, line 338 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0BAIR6DwbIMJCGOLam:-L0G-1jIVZnKrLM-uuqk:b-9g8amf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72d9cf0bacccf752dbe5de9fc367614c857f460/pkg/sql/distinct.go#L338)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

:+1: done
</blockquote></details>

As an aside, it's interesting that here we know it's a real key. We define strong keys as weak keys on non-null columns, but that could be loosened to include this case, where columns can have NULLs but we still don't have two rows that are equal (but where NULL equals NULL) on these columns.

---

*[pkg/sql/plan_physical_props.go, line 125 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0G05xqNdJJMG-yZWLv:-L0G05xqNdJJMG-yZWLw:b5uagga) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffec33995ca68a379b87ac8e6a87e78a149dcfcd/pkg/sql/plan_physical_props.go#L125)):*
> ```Go
> 	}
> 
> 	// We can potentially reduce ordering since we've added some weak keys.
> ```

Oh, I think this is an omission in `addWeakKey`, we should do this there (or every caller needs to reduce)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-:-L0G0vwHGK9-wL2s4OD3:bnfp4nl)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Review status: 0 of 24 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/distinct.go, line 338 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0BAIR6DwbIMJCGOLam:-L0GMj2ryVBRC0IxMHBT:bc0ad3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72d9cf0bacccf752dbe5de9fc367614c857f460/pkg/sql/distinct.go#L338)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

As an aside, it's interesting that here we know it's a real key. We define strong keys as weak keys on non-null columns, but that could be loosened to include this case, where columns can have NULLs but we still don't have two rows that are equal (but where NULL equals NULL) on these columns.
</blockquote></details>

What do you think about a hint for this type of situation for every "weak key" in physical props? I can't think of a query where `DISTINCT`'s key property is useful for planning, but it might help us further down the road.

---

*[pkg/sql/plan_physical_props.go, line 125 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0G05xqNdJJMG-yZWLv:-L0GgXhbanHn71oXuDj0:bouqrgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffec33995ca68a379b87ac8e6a87e78a149dcfcd/pkg/sql/plan_physical_props.go#L125)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Oh, I think this is an omission in `addWeakKey`, we should do this there (or every caller needs to reduce)
</blockquote></details>

Hrm it looks like it is an omission. Fixed :+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 14 files at r1, 5 of 17 files at r2, 3 of 8 files at r3, 7 of 9 files at r4, 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/distinct.go, line 338 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-L0BAIR6DwbIMJCGOLam:-L0Gi1CeRvK9Op_enNxO:bj7x9jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72d9cf0bacccf752dbe5de9fc367614c857f460/pkg/sql/distinct.go#L338)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

What do you think about a hint for this type of situation for every "weak key" in physical props? I can't think of a query where `DISTINCT`'s key property is useful for planning, but it might help us further down the road.
</blockquote></details>

I am not sure. We may want to store a separate list of strong keys (and move them from weak to strong when columns become non-null). I am not yet convinced it's worth it though

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20463#-:-L0GiNyDfVQX9j-bFqjH:bnfp4nl)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Forgot to cc @jseldess: this will require some additional documentation for 2.0 - we can sync up sometime if you want.
